File: frontend/src/lib/ddos.ts
/**
 * AIVAR RedForge — DDoS Defense Center API client (M19).
 *
 * Every call is tenant-scoped by the backend from the bearer token;
 * organization_id is never sent from the client.
 *
 * All data returned from these endpoints is derived from canonical
 * telemetry_events aggregation. No synthetic or fabricated values.
 */

import { api } from "@/lib/api";

// ── Protected Resources ───────────────────────────────────────────────────────

export interface ProtectedResource {
  id: string;
  name: string;
  description: string;
  scope_type: string;
  scope_value: string | null;
  criticality: string;
  monitored_ports: number[] | null;
  monitoring_enabled: boolean;
  created_by: string;
  created_at: string;
  updated_at: string;
}

export interface DetectionPolicy {
  id: string;
  resource_id: string;
  enabled: boolean;
  profile: string;
  static_bps_threshold: number | null;
  static_pps_threshold: number | null;
  static_fps_threshold: number | null;
  window_seconds: number;
  min_breach_windows: number;
  quiet_period_windows: number;
  mitigation_mode: string;
  suppression_windows: unknown[] | null;
  created_at: string;
  updated_at: string;
}

export function listProtectedResources(): Promise<ProtectedResource[]> {
  return api.get<ProtectedResource[]>("/api/v1/ddos/resources");
}

export function getProtectedResource(id: string): Promise<ProtectedResource> {
  return api.get<ProtectedResource>(`/api/v1/ddos/resources/${id}`);
}

export function createProtectedResource(data: {
  name: string;
  description?: string;
  scope_type?: string;
  scope_value?: string;
  criticality?: string;
  monitored_ports?: number[];
}): Promise<ProtectedResource> {
  return api.post<ProtectedResource>("/api/v1/ddos/resources", data);
}

export function deleteProtectedResource(id: string): Promise<void> {
  return api.delete<void>(`/api/v1/ddos/resources/${id}`);
}

export function getDetectionPolicy(resourceId: string): Promise<DetectionPolicy> {
  return api.get<DetectionPolicy>(`/api/v1/ddos/resources/${resourceId}/policy`);
}

// ── Incidents ─────────────────────────────────────────────────────────────────

export interface DDoSIncident {
  id: string;
  resource_id: string;
  resource_name: string;
  status: string;
  severity: string;
  classification: string;
  first_detected_at: string;
  last_updated_at: string;
  peak_at: string | null;
  resolved_at: string | null;
  peak_bytes_per_second: number | null;
  peak_packets_per_second: number | null;
  peak_flows_per_second: number | null;
  peak_unique_src_ips: number | null;
  peak_deviation_multiplier: number | null;
  consecutive_quiet_windows: number;
  opening_evidence: Record<string, unknown>;
  latest_evidence: Record<string, unknown>;
}

export interface IncidentTimelineEvent {
  id: string;
  event_type: string;
  occurred_at: string;
  description: string;
  payload: Record<string, unknown>;
  actor_id: string | null;
}

export function listIncidents(params?: {
  status?: string[];
  resource_id?: string;
  limit?: number;
  offset?: number;
}): Promise<DDoSIncident[]> {
  const qs = new URLSearchParams();
  if (params?.status) params.status.forEach((s) => qs.append("status", s));
  if (params?.resource_id) qs.set("resource_id", params.resource_id);
  if (params?.limit) qs.set("limit", String(params.limit));
  if (params?.offset) qs.set("offset", String(params.offset));
  const q = qs.toString();
  return api.get<DDoSIncident[]>(`/api/v1/ddos/incidents${q ? `?${q}` : ""}`);
}

export function listActiveIncidents(): Promise<DDoSIncident[]> {
  return api.get<DDoSIncident[]>("/api/v1/ddos/incidents/active");
}

export function getIncident(id: string): Promise<DDoSIncident> {
  return api.get<DDoSIncident>(`/api/v1/ddos/incidents/${id}`);
}

export function getIncidentTimeline(id: string): Promise<IncidentTimelineEvent[]> {
  return api.get<IncidentTimelineEvent[]>(`/api/v1/ddos/incidents/${id}/timeline`);
}

export function closeIncident(id: string): Promise<DDoSIncident> {
  return api.post<DDoSIncident>(`/api/v1/ddos/incidents/${id}/close`, { reason: "" });
}

// ── Mitigation ────────────────────────────────────────────────────────────────

export interface MitigationRecommendation {
  id: string;
  incident_id: string;
  recommendation_type: string;
  description: string;
  recommendation_detail: Record<string, unknown>;
  approval_status: string;
  approved_by: string | null;
  approved_at: string | null;
  rejected_by: string | null;
  rejection_reason: string | null;
  execution_status: string;
  executed_at: string | null;
  provider_type: string | null;
  expires_at: string | null;
  created_at: string;
}

export function listPendingRecommendations(): Promise<MitigationRecommendation[]> {
  return api.get<MitigationRecommendation[]>("/api/v1/ddos/mitigation/pending");
}

export function listIncidentRecommendations(incidentId: string): Promise<MitigationRecommendation[]> {
  return api.get<MitigationRecommendation[]>(`/api/v1/ddos/incidents/${incidentId}/mitigation`);
}

export function approveRecommendation(recId: string): Promise<MitigationRecommendation> {
  return api.post<MitigationRecommendation>(`/api/v1/ddos/mitigation/${recId}/approve`, {});
}

export function rejectRecommendation(recId: string, reason: string): Promise<MitigationRecommendation> {
  return api.post<MitigationRecommendation>(`/api/v1/ddos/mitigation/${recId}/reject`, { reason });
}

// ── Traffic Windows ───────────────────────────────────────────────────────────

export interface ObservationWindow {
  id: string;
  resource_id: string;
  window_start_ts: string;
  window_end_ts: string;
  event_count: number;
  total_bytes_in: number | null;
  total_bytes_out: number | null;
  unique_src_ips: number;
  protocol_counts: Record<string, number>;
  alert_count: number;
  detection_fired: boolean;
  severity: string | null;
  classification: string | null;
  incident_id: string | null;
}

export function listTrafficWindows(resourceId: string, hours?: number): Promise<ObservationWindow[]> {
  const qs = new URLSearchParams({ resource_id: resourceId });
  if (hours) qs.set("hours", String(hours));
  return api.get<ObservationWindow[]>(`/api/v1/ddos/traffic/windows?${qs}`);
}

// ── Posture ───────────────────────────────────────────────────────────────────

export interface DDoSPosture {
  protected_resource_count: number;
  active_incident_count: number;
  pending_recommendation_count: number;
  active_incident_severity_counts: Record<string, number>;
  active_incidents: Array<{
    id: string;
    resource_id: string;
    resource_name: string;
    status: string;
    severity: string;
    classification: string;
    first_detected_at: string;
  }>;
}

export function getDDoSPosture(): Promise<DDoSPosture> {
  return api.get<DDoSPosture>("/api/v1/ddos/posture");
}

// ── Helpers ───────────────────────────────────────────────────────────────────

export function severityColor(severity: string): string {
  switch (severity?.toUpperCase()) {
    case "CRITICAL": return "text-red-400 border-red-800 bg-red-950";
    case "HIGH":     return "text-orange-400 border-orange-800 bg-orange-950";
    case "MEDIUM":   return "text-yellow-400 border-yellow-800 bg-yellow-950";
    case "LOW":      return "text-blue-400 border-blue-800 bg-blue-950";
    default:         return "text-gray-400 border-gray-700 bg-gray-800";
  }
}

export function statusColor(status: string): string {
  switch (status?.toUpperCase()) {
    case "DETECTED":   return "text-red-300 border-red-900 bg-red-950";
    case "ACTIVE":     return "text-orange-300 border-orange-900 bg-orange-950";
    case "ESCALATED":  return "text-red-400 border-red-800 bg-red-900";
    case "MITIGATING": return "text-yellow-300 border-yellow-900 bg-yellow-950";
    case "MONITORING": return "text-blue-300 border-blue-900 bg-blue-950";
    case "RESOLVED":   return "text-emerald-300 border-emerald-900 bg-emerald-950";
    case "CLOSED":     return "text-gray-400 border-gray-700 bg-gray-800";
    default:           return "text-gray-400 border-gray-700 bg-gray-800";
  }
}

export function formatBps(bps: number | null): string {
  if (bps == null) return "—";
  if (bps >= 1e9) return `${(bps / 1e9).toFixed(2)} Gbps`;
  if (bps >= 1e6) return `${(bps / 1e6).toFixed(2)} Mbps`;
  if (bps >= 1e3) return `${(bps / 1e3).toFixed(1)} Kbps`;
  return `${bps.toFixed(0)} bps`;
}

export function formatPps(pps: number | null): string {
  if (pps == null) return "—";
  if (pps >= 1e6) return `${(pps / 1e6).toFixed(2)} Mpps`;
  if (pps >= 1e3) return `${(pps / 1e3).toFixed(1)} Kpps`;
  return `${pps.toFixed(0)} pps`;
}

File: backend/src/redforge/api/v1/ddos.py
"""DDoS Defense Center REST API — M19.

All routes are tenant-scoped; organization_id is derived from JWT,
never accepted from the client.

Permission gates:
  DDOS_READ              — all read endpoints (GET)
  DDOS_MANAGE            — resource/policy management (POST, PUT, DELETE)
  DDOS_MITIGATION_APPROVE — mitigation approve/reject actions
"""

from __future__ import annotations

from dataclasses import asdict
from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, Query
from pydantic import BaseModel, Field

from redforge.api.security import TenantContext, require_permission
from redforge.domain.identity.value_objects import Permission

router = APIRouter(prefix="/ddos")


# ── Pydantic request/response models ─────────────────────────────────────────


class CreateResourceRequest(BaseModel):
    name: str = Field(..., min_length=1, max_length=200)
    description: str = Field("", max_length=1000)
    scope_type: str = Field("any")
    scope_value: str | None = None
    criticality: str = Field("MEDIUM")
    monitored_ports: list[int] | None = None


class UpsertPolicyRequest(BaseModel):
    enabled: bool = True
    profile: str = Field("BALANCED")
    static_bps_threshold: float | None = None
    static_pps_threshold: float | None = None
    static_fps_threshold: float | None = None
    window_seconds: int = Field(60, ge=30, le=3600)
    min_breach_windows: int = Field(1, ge=1, le=10)
    quiet_period_windows: int = Field(3, ge=1, le=20)
    mitigation_mode: str = Field("RECOMMEND_ONLY")


class CloseIncidentRequest(BaseModel):
    reason: str = Field("", max_length=500)


class MitigationActionRequest(BaseModel):
    reason: str = Field("", max_length=500)


# ── Dependency helpers ────────────────────────────────────────────────────────


def _get_session_factory() -> Any:
    from redforge.api.dependencies import _session_factory
    return _session_factory()


async def _resource_svc(session_factory: Any = Depends(_get_session_factory)) -> Any:

    from redforge.application.ddos.resource_service import DDoSResourceService

    async with session_factory() as session, session.begin():
        yield DDoSResourceService(session)


async def _incident_svc(session_factory: Any = Depends(_get_session_factory)) -> Any:
    from redforge.application.ddos.incident_service import DDoSIncidentService

    async with session_factory() as session, session.begin():
        yield DDoSIncidentService(session)


# ── Protected Resource Endpoints ──────────────────────────────────────────────


@router.get("/resources")
async def list_protected_resources(
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    session_factory: Any = Depends(_get_session_factory),
) -> list[dict[str, Any]]:
    from redforge.application.ddos.resource_service import DDoSResourceService

    async with session_factory() as session, session.begin():
        svc = DDoSResourceService(session)
        resources = await svc.list_resources(tenant.organization_id)
    return [asdict(r) for r in resources]


@router.post("/resources", status_code=201)
async def create_protected_resource(
    body: CreateResourceRequest,
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_MANAGE))],
    session_factory: Any = Depends(_get_session_factory),
) -> dict[str, Any]:
    from redforge.application.ddos.resource_service import DDoSResourceService

    async with session_factory() as session, session.begin():
        svc = DDoSResourceService(session)
        resource = await svc.create_resource(
            organization_id=tenant.organization_id,
            actor_id=tenant.user_id,
            name=body.name,
            description=body.description,
            scope_type=body.scope_type,
            scope_value=body.scope_value,
            criticality=body.criticality,
            monitored_ports=body.monitored_ports,
        )
    return asdict(resource)


@router.get("/resources/{resource_id}")
async def get_protected_resource(
    resource_id: str,
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    session_factory: Any = Depends(_get_session_factory),
) -> dict[str, Any]:
    from redforge.application.ddos.resource_service import DDoSResourceService

    async with session_factory() as session, session.begin():
        svc = DDoSResourceService(session)
        resource = await svc.get_resource(tenant.organization_id, resource_id)
    if resource is None:
        raise HTTPException(status_code=404, detail="Protected resource not found")
    return asdict(resource)


@router.delete("/resources/{resource_id}", status_code=204)
async def delete_protected_resource(
    resource_id: str,
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_MANAGE))],
    session_factory: Any = Depends(_get_session_factory),
) -> None:
    from redforge.application.ddos.resource_service import DDoSResourceService

    async with session_factory() as session, session.begin():
        svc = DDoSResourceService(session)
        deleted = await svc.delete_resource(tenant.organization_id, resource_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Protected resource not found")


# ── Detection Policy Endpoints ────────────────────────────────────────────────


@router.get("/resources/{resource_id}/policy")
async def get_detection_policy(
    resource_id: str,
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    session_factory: Any = Depends(_get_session_factory),
) -> dict[str, Any]:
    from redforge.application.ddos.resource_service import DDoSResourceService

    async with session_factory() as session, session.begin():
        svc = DDoSResourceService(session)
        policy = await svc.get_policy(tenant.organization_id, resource_id)
    if policy is None:
        raise HTTPException(status_code=404, detail="Detection policy not found")
    return asdict(policy)


@router.put("/resources/{resource_id}/policy", status_code=200)
async def upsert_detection_policy(
    resource_id: str,
    body: UpsertPolicyRequest,
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_MANAGE))],
    session_factory: Any = Depends(_get_session_factory),
) -> dict[str, Any]:
    from redforge.application.ddos.resource_service import DDoSResourceService

    async with session_factory() as session, session.begin():
        svc = DDoSResourceService(session)
        # Verify resource exists and belongs to org
        resource = await svc.get_resource(tenant.organization_id, resource_id)
        if resource is None:
            raise HTTPException(status_code=404, detail="Protected resource not found")
        policy = await svc.upsert_policy(
            organization_id=tenant.organization_id,
            resource_id=resource_id,
            actor_id=tenant.user_id,
            enabled=body.enabled,
            profile=body.profile,
            static_bps_threshold=body.static_bps_threshold,
            static_pps_threshold=body.static_pps_threshold,
            static_fps_threshold=body.static_fps_threshold,
            window_seconds=body.window_seconds,
            min_breach_windows=body.min_breach_windows,
            quiet_period_windows=body.quiet_period_windows,
            mitigation_mode=body.mitigation_mode,
        )
    return asdict(policy)


# ── Incident Endpoints ────────────────────────────────────────────────────────


@router.get("/incidents")
async def list_incidents(
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    status: list[str] = Query(default=[]),
    resource_id: str | None = Query(default=None),
    limit: int = Query(default=100, ge=1, le=500),
    offset: int = Query(default=0, ge=0),
    session_factory: Any = Depends(_get_session_factory),
) -> list[dict[str, Any]]:
    from redforge.application.ddos.incident_service import DDoSIncidentService

    async with session_factory() as session, session.begin():
        svc = DDoSIncidentService(session)
        incidents = await svc.list_incidents(
            organization_id=tenant.organization_id,
            status_filter=status or None,
            resource_id=resource_id,
            limit=limit,
            offset=offset,
        )
    return [asdict(i) for i in incidents]


@router.get("/incidents/active")
async def list_active_incidents(
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    session_factory: Any = Depends(_get_session_factory),
) -> list[dict[str, Any]]:
    from redforge.application.ddos.incident_service import DDoSIncidentService

    async with session_factory() as session, session.begin():
        svc = DDoSIncidentService(session)
        incidents = await svc.list_active_incidents(tenant.organization_id)
    return [asdict(i) for i in incidents]


@router.get("/incidents/{incident_id}")
async def get_incident(
    incident_id: str,
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    session_factory: Any = Depends(_get_session_factory),
) -> dict[str, Any]:
    from redforge.application.ddos.incident_service import DDoSIncidentService

    async with session_factory() as session, session.begin():
        svc = DDoSIncidentService(session)
        incident = await svc.get_incident(tenant.organization_id, incident_id)
    if incident is None:
        raise HTTPException(status_code=404, detail="Incident not found")
    return asdict(incident)


@router.post("/incidents/{incident_id}/close", status_code=200)
async def close_incident(
    incident_id: str,
    body: CloseIncidentRequest,
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_MANAGE))],
    session_factory: Any = Depends(_get_session_factory),
) -> dict[str, Any]:
    from redforge.application.ddos.incident_service import DDoSIncidentService

    async with session_factory() as session, session.begin():
        svc = DDoSIncidentService(session)
        incident = await svc.close_incident(
            organization_id=tenant.organization_id,
            incident_id=incident_id,
            actor_id=tenant.user_id,
        )
    if incident is None:
        raise HTTPException(status_code=404, detail="Incident not found")
    return asdict(incident)


@router.get("/incidents/{incident_id}/timeline")
async def get_incident_timeline(
    incident_id: str,
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    session_factory: Any = Depends(_get_session_factory),
) -> list[dict[str, Any]]:
    from redforge.application.ddos.incident_service import DDoSIncidentService

    async with session_factory() as session, session.begin():
        svc = DDoSIncidentService(session)
        # Verify incident belongs to org
        incident = await svc.get_incident(tenant.organization_id, incident_id)
        if incident is None:
            raise HTTPException(status_code=404, detail="Incident not found")
        timeline = await svc.get_timeline(tenant.organization_id, incident_id)
    return [asdict(e) for e in timeline]


@router.get("/incidents/{incident_id}/mitigation")
async def list_incident_recommendations(
    incident_id: str,
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    session_factory: Any = Depends(_get_session_factory),
) -> list[dict[str, Any]]:
    from redforge.application.ddos.incident_service import DDoSIncidentService

    async with session_factory() as session, session.begin():
        svc = DDoSIncidentService(session)
        recs = await svc.list_recommendations(tenant.organization_id, incident_id)
    return [asdict(r) for r in recs]


# ── Mitigation Recommendation Approval ───────────────────────────────────────


@router.get("/mitigation/pending")
async def list_pending_recommendations(
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    session_factory: Any = Depends(_get_session_factory),
) -> list[dict[str, Any]]:
    from redforge.application.ddos.incident_service import DDoSIncidentService

    async with session_factory() as session, session.begin():
        svc = DDoSIncidentService(session)
        recs = await svc.list_pending_recommendations(tenant.organization_id)
    return [asdict(r) for r in recs]


@router.post("/mitigation/{recommendation_id}/approve", status_code=200)
async def approve_recommendation(
    recommendation_id: str,
    tenant: Annotated[
        TenantContext, Depends(require_permission(Permission.DDOS_MITIGATION_APPROVE))
    ],
    session_factory: Any = Depends(_get_session_factory),
) -> dict[str, Any]:
    from redforge.application.ddos.incident_service import DDoSIncidentService

    async with session_factory() as session, session.begin():
        svc = DDoSIncidentService(session)
        rec = await svc.approve_recommendation(
            organization_id=tenant.organization_id,
            rec_id=recommendation_id,
            actor_id=tenant.user_id,
        )
    if rec is None:
        raise HTTPException(status_code=404, detail="Recommendation not found")
    return asdict(rec)


@router.post("/mitigation/{recommendation_id}/reject", status_code=200)
async def reject_recommendation(
    recommendation_id: str,
    body: MitigationActionRequest,
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_MANAGE))],
    session_factory: Any = Depends(_get_session_factory),
) -> dict[str, Any]:
    from redforge.application.ddos.incident_service import DDoSIncidentService

    async with session_factory() as session, session.begin():
        svc = DDoSIncidentService(session)
        rec = await svc.reject_recommendation(
            organization_id=tenant.organization_id,
            rec_id=recommendation_id,
            actor_id=tenant.user_id,
            reason=body.reason,
        )
    if rec is None:
        raise HTTPException(status_code=404, detail="Recommendation not found")
    return asdict(rec)


# ── Traffic / Window Analytics ────────────────────────────────────────────────


@router.get("/traffic/windows")
async def list_observation_windows(
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    resource_id: str = Query(...),
    hours: int = Query(default=1, ge=1, le=72),
    session_factory: Any = Depends(_get_session_factory),
) -> list[dict[str, Any]]:
    from datetime import UTC, timedelta

    from redforge.infrastructure.database.repositories.ddos.window_repository import (
        SqlAlchemyObservationWindowRepository,
    )
    since = __import__("datetime").datetime.now(UTC) - timedelta(hours=hours)

    async with session_factory() as session, session.begin():
        repo = SqlAlchemyObservationWindowRepository(session)
        windows = await repo.list_for_resource(
            organization_id=tenant.organization_id,
            resource_id=resource_id,
            since=since,
            limit=min(hours * 60 + 10, 5000),
        )
    return [
        {
            "id": w.id,
            "resource_id": w.resource_id,
            "window_start_ts": w.window_start_ts.isoformat(),
            "window_end_ts": w.window_end_ts.isoformat(),
            "event_count": w.event_count,
            "total_bytes_in": w.total_bytes_in,
            "total_bytes_out": w.total_bytes_out,
            "unique_src_ips": w.unique_src_ips,
            "protocol_counts": w.protocol_counts,
            "alert_count": w.alert_count,
            "detection_fired": w.detection_fired,
            "severity": w.severity,
            "classification": w.classification,
            "incident_id": w.incident_id,
        }
        for w in windows
    ]


# ── Posture Summary ───────────────────────────────────────────────────────────


@router.get("/posture")
async def get_ddos_posture(
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    session_factory: Any = Depends(_get_session_factory),
) -> dict[str, Any]:
    """Return org-level DDoS posture: active incidents, resource count, recent windows."""
    from redforge.application.ddos.incident_service import DDoSIncidentService
    from redforge.application.ddos.resource_service import DDoSResourceService

    async with session_factory() as session, session.begin():
        res_svc = DDoSResourceService(session)
        inc_svc = DDoSIncidentService(session)

        resources = await res_svc.list_resources(tenant.organization_id)
        active_incidents = await inc_svc.list_active_incidents(tenant.organization_id)
        pending_recs = await inc_svc.list_pending_recommendations(tenant.organization_id)

    severity_counts: dict[str, int] = {}
    for inc in active_incidents:
        severity_counts[inc.severity] = severity_counts.get(inc.severity, 0) + 1

    return {
        "protected_resource_count": len(resources),
        "active_incident_count": len(active_incidents),
        "pending_recommendation_count": len(pending_recs),
        "active_incident_severity_counts": severity_counts,
        "active_incidents": [
            {
                "id": i.id,
                "resource_id": i.resource_id,
                "resource_name": i.resource_name,
                "status": i.status,
                "severity": i.severity,
                "classification": i.classification,
                "first_detected_at": i.first_detected_at,
            }
            for i in active_incidents
        ],
    }


# ── Worker Health ─────────────────────────────────────────────────────────────


@router.get("/worker/health")
async def get_worker_health(
    tenant: Annotated[TenantContext, Depends(require_permission(Permission.DDOS_READ))],
    request: Any = None,
) -> dict[str, Any]:
    """Return DDoS detection worker stats from app state."""

    # Worker is stored on app.state by the startup lifecycle
    # Returns empty stats if worker not wired (test/development)
    worker = getattr(request.app.state, "ddos_detection_worker", None) if request else None
    if worker is None:
        return {"status": "not_configured", "stats": {}}
    return {"status": "running" if worker._running else "stopped", "stats": worker.stats()}
